github/resource_github_actions_secret.go

```diff
@@ -0,0 +1,157 @@
+package github
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"github.com/google/go-github/v29/github"
+	"github.com/hashicorp/terraform/helper/schema"
+	"golang.org/x/crypto/nacl/box"
+	"log"
+)
+
+func resourceGithubActionsSecret() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGithubActionsSecretCreateOrUpdate,
+		Read:   resourceGithubActionsSecretRead,
+		Update: resourceGithubActionsSecretCreateOrUpdate,
+		Delete: resourceGithubActionsSecretDelete,
+
+		Schema: map[string]*schema.Schema{
+			"owner": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"repository": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"secret_name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"plaintext_value": {
+				Type:      schema.TypeString,
+				Required:  true,
+				Sensitive: true,
+			},
+		},
+	}
+}
+
+func resourceGithubActionsSecretCreateOrUpdate(d *schema.ResourceData, meta interface{}) error {
+	err := checkOrganization(meta)
+	if err != nil {
+		return err
+	}
+
+	client := meta.(*Organization).client
+	ctx := context.Background()
+
+	repo := d.Get("repository").(string)
+	owner := d.Get("owner").(string)
+	secretName := d.Get("secret_name").(string)
+	plaintextValue := d.Get("plaintext_value").(string)
+
+	keyId, publicKey, err := getPublicKeyDetails(owner, repo, meta)
+	if err != nil {
+		return err
+	}
+
+	encryptedText, err := encryptPlaintext(plaintextValue, publicKey)
+	if err != nil {
+		return err
+	}
+
+	// Create an EncryptedSecret and encrypt the plaintext value into it
+	eSecret := &github.EncryptedSecret{
+		Name:           secretName,
+		KeyID:          keyId,
+		EncryptedValue: base64.StdEncoding.EncodeToString(encryptedText),
+	}
+
+	_, err = client.Actions.CreateOrUpdateSecret(ctx, owner, repo, eSecret)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(fmt.Sprintf("%s/%s:%s", owner, repo, secretName))
+
+	return resourceGithubActionsSecretRead(d, meta)
+}
+
+func resourceGithubActionsSecretRead(d *schema.ResourceData, meta interface{}) error {
+	err := checkOrganization(meta)
+	if err != nil {
+		return err
+	}
+
+	client := meta.(*Organization).client
+	ctx := context.Background()
+
+	repo := d.Get("repository").(string)
+	owner := d.Get("owner").(string)
+	secretName := d.Get("secret_name").(string)
+
+	secret, _, err := client.Actions.GetSecret(ctx, owner, repo, secretName)
+	if err != nil || secret.Name != secretName {
+		d.SetId("")
+		return err
+	}
+
```

Comment:
How come there are no `d.Set()` method calls? I think the secret information could be stored in state. The `UpdatedAt` value from `github.Secret` could be used to determine if the value has changed in GitHub, for example, if `UpdatedAt` stored in state differs from value you got in `GetSecret` response, it means someone has updated the value outside of Terraform, and we probably want to sync it up (reapply value defined in Terraform).

Reply:
I wasn't aware that the `d.Set()` functions were required to pass things to state. My understanding was that the existing `d` values passed in would get persisted.
I fully agree with the use of `UpdatedAt` - would you recommend saving the full state of the `github.Secret` ?

Reply:
I think I would store the `repo` and `name` as par of Id, then store `plaintext_value` as it is and lastly store `updated_at` as computed getting the value from create/update operation.

Reply:
Sounds like a solid plan - I will also include `created_at` since we have it available and might as well make it available